### ISSUE TITLE
Agent prompt token bloat: measure, slim, and modularize

### ISSUE THREAD

**Issue by alteixeira20** [pewdiepie-archdaemon/odysseus#2750]
## Area

Chat / Agent

## Problem or Motivation

Agent-mode system prompts can approach roughly 10K tokens, which crowds context on smaller/self-hosted models with 4K/8K/16K context windows and can reduce reliability before the user request, conversation context, tool results, or documents are even considered.

This issue tracks the project-side prompt-bloat work. It is separate from the “which model should I run?” problem. Model benchmarking, Cookbook recommendations, and a possible finetune can still happen later, but the first step should be to measure and reduce unnecessary default prompt load.

This is a parent roadmap/tracking issue, not a single large implementation request. Work should happen through small follow-up PRs.

Current audit-based findings:

- Agent prompt assembly appears to be concentrated mainly in `src/agent_loop.py`.
- Per-tool guidance and native function-call schemas appear to already be partially filtered by tool selection, so not all large prompt components should be treated as always-on.
- Some expensive areas still appear to be always loaded or insufficiently filtered:
  - MCP prose descriptions;
  - workflow/rules blocks;
  - native-agent rules;
  - integrations prompt text;
  - RAG-down fallback behavior;
  - some tool/group-specific rules embedded in general guidance.
- MCP prose descriptions appear to be especially worth checking early: they seem to be appended for all connected MCP tools instead of being filtered to the selected/relevant tool set. Native schemas are already filtered, so MCP prose likely should follow the same principle.
- Tool descriptions also appear to exist in multiple places (`TOOL_SECTIONS`, `BUILTIN_TOOL_DESCRIPTIONS`, `FUNCTION_TOOL_SCHEMAS`), which creates drift and maintenance cost.

Right now we do not have a per-request prompt token breakdown, so we cannot accurately answer:

- how large the normal assembled agent prompt is;
- how often the heavy/fallback path is used;
- how much comes from rules vs. tool guidance vs. schemas vs. MCP vs. integrations;
- how many tools were selected vs. actually sent;
- which optimization area is worth doing first.

That measurement gap should be fixed before making behavior-sensitive prompt cuts.

## Proposed Solution

Measure first, then reduce prompt size through small, reversible PRs.

### Goals

- Add visibility into assembled agent prompt token usage.
- Break prompt cost down by component.
- Reduce default prompt size without removing tools or capabilities.
- Modularize guidance so task/tool-specific instructions are loaded just in time.
- Improve reliability on smaller/self-hosted models.
- Add enough eval coverage to make prompt cuts measurable rather than vibes-based.
- Keep a future finetune possible, but make it evidence-driven.

### Non-goals

- No broad rewrite of the agent loop.
- No removal of tools, capabilities, or MCP support.
- No immediate finetune work in this issue.
- No behavior-sensitive slimming before measurement/evals.
- No change to the normal non-agent chat path unless measurement shows it is relevant.
- No large all-in-one PR.

### Phase 0 — Measurement

First PR should be measurement-only.

Add diagnostics for assembled agent prompt token usage:

- total assembled prompt estimate;
- per-component estimate;
- selected tool count vs. sent tool count;
- whether the path used normal tool selection or fallback;
- no sensitive user content in logs;
- no prompt wording change;
- no tool-selection change;
- no behavior change.

Suggested first PR title:

`chore(agent): add assembled-prompt token breakdown diagnostics`

This should establish the baseline for every later cut.

### Phase 1 — Safe low-risk slimming

After measurement exists, make small behavior-preserving reductions.

Candidate PRs:

1. **Filter MCP prose descriptions by selected/relevant tools**

   Only include MCP prose descriptions for selected/relevant MCP tools, instead of appending prose for all connected MCP tools.

2. **Slim the RAG-down fallback**

   Avoid falling back to the full agent system prompt when tool selection/RAG is unavailable. Use a smaller core or keyword-selected set instead.

3. **Gate integrations prompt text**

   Only inject integrations guidance when integration/API-call tools are relevant to the request.

4. **Gate low-value trailing hints for small-context models**

   Avoid spending tokens on generic “other tools are available” style hints for small-context models unless they are proven useful.

### Phase 2 — Just-in-time tool guidance

Once measurement and basic eval coverage exist, split large always-on rule blocks into smaller fragments:

- core rules always loaded;
- cookbook-specific rules only when cookbook tools are selected;
- email-specific rules only when email tools are selected;
- calendar/task-specific rules only when those tools are selected;
- document/research-specific rules only when those tools are selected;
- memory-specific rules only when memory tools are selected;
- MCP-specific guidance only when relevant MCP tools are selected.

This phase is more behavior-sensitive and should be guarded by evals.

### Phase 3 — Eval suite and model benchmarking

Build a small replay/eval suite while prompt slimming work progresses.

Suggested coverage:

- simple chat without tools;
- single-tool calls;
- multi-tool agent flows;
- notes/reminders;
- calendar/task scheduling;
- memory edit/audit;
- document workflows;
- deep research;
- cookbook/model-selection decisions;
- MCP tool selection;
- safety and prompt-injection-sensitive tool boundaries.

Useful metrics:

- assembled prompt tokens;
- tool-selection accuracy;
- task completion;
- regression rate on a strong reference model;
- smaller-model success rate;
- latency / prefill impact;
- safety fixture pass/fail.

This can later support a public model scoreboard for the “which model should I run?” question.

### Phase 4 — Optional finetune decision

Keep finetuning on the table, but not as the first implementation step.

Only consider a dedicated Odysseus finetune if, after measurement, slimming, modular guidance, and benchmarking:

- smaller models still fail important task surfaces;
- failures appear model-capability-related rather than prompt-architecture-related;
- evals can prove whether the finetune improves behavior;
- trace/sample data is opt-in, redacted, and reproducible.

### Lightweight sample collection

Sample collection can run lightly in parallel, but it should not become the immediate priority.

Initial purpose should be eval data, not training data. That way the work remains useful even if no finetune is ever trained.

Any trace collection should be:

- opt-in;
- off by default;
- local-first;
- redacted;
- careful around emails, contacts, paths, memory content, API keys, tokens, and personal data.

### Finetune first

A dedicated Odysseus finetune may be useful later, but it should not be the first step.

Training first risks overfitting to today’s prompt wording, tool names, and tool layout. It could also create a retrain-on-every-tool-change treadmill before we know how much of the problem can be fixed by prompt architecture.

### Model recommendations only

Better Cookbook recommendations and a public benchmark/scoreboard would help self-hosted users choose existing 1B–14B models. That is valuable, but it does not replace the need to reduce unnecessary default prompt load.

### Broad agent-loop rewrite

Not desirable here. This should be incremental: measurement first, then small PRs.

## Prior Art / Related Issues

- ROADMAP.md: agent prompt/context bloat.
- ROADMAP.md: skill/tool prompt-injection audit.
- Related discussion: prompt bloat, small models, model benchmarking, sample collection, and possible finetuning.
- Relevant code areas:
  - `src/agent_loop.py`
  - `src/tool_index.py`
  - `src/tool_schemas.py`
  - `src/mcp_manager.py`
  - `src/model_context.py`

## Are you willing to implement this?

Yes — I can open a PR.

**Comment by undergroundrap** [comment]
## Proposal: measurement-only agent prompt budget diagnostic

Following the discussion in #2724, I’d like to propose a small first PR for the prompt-bloat roadmap: a read-only diagnostic that reports where the assembled agent prompt budget is going.

The goal is to create a baseline for later slimming/modularization PRs without changing behavior.

## Scope

Measurement only:

- no prompt wording changes;
- no routing changes;
- no tool-selection changes;
- no model behavior changes.

## Proposed breakdown

The diagnostic should build the same assembled agent prompt shape the runtime would send, then report section-level size data for:

- base agent preamble / rules;
- API-agent additions if present;
- builtin tool sections;
- MCP tool descriptions;
- skill index / skill snippets;
- active document or active-context injection;
- integration-specific prompt additions;
- date/time/user-context preface;
- top largest sections/tools by estimated token count.

## Proposed fields

For each section:

- `name`;
- `category`;
- `char_count`;
- `estimated_tokens`, using the existing estimator where available;
- `percent_of_total`;
- optional `item_count` for grouped sections such as tools, MCP tools, or skills.

For largest-section output:

- section/tool name;
- category;
- char count;
- estimated token count;
- percent of total.

## Possible output shape

```json
{
  "total": {
    "char_count": 42150,
    "estimated_tokens": 12645
  },
  "sections": [
    {
      "name": "base_agent_rules",
      "category": "base",
      "char_count": 8200,
      "estimated_tokens": 2460,
      "percent_of_total": 19.5
    },
    {
      "name": "builtin_tool_sections",
      "category": "tools",
      "char_count": 17400,
      "estimated_tokens": 5220,
      "percent_of_total": 41.3,
      "item_count": 38
    }
  ],
  "largest": [
    {
      "name": "serve_model",
      "category": "builtin_tool",
      "char_count": 2100,
      "estimated_tokens": 630,
      "percent_of_total": 5.0
    }
  ]
}
```

**Comment by undergroundrap** [comment]
One small addendum on implementation shape:

I’d keep the first PR small and non-invasive:

1. Add a helper that assembles prompt-budget sections and computes char/token totals.
2. Reuse the existing token estimator where practical; if a section cannot be estimated cleanly, keep char counts rather than forcing a refactor.
3. Add focused tests for:
   - total equals sum of sections;
   - percentages are stable;
   - largest sections sort descending;
   - empty optional sections are handled cleanly.
4. Expose the diagnostic in the least invasive place maintainers prefer: CLI/dev-only route/log helper/test fixture.

I’m intentionally not proposing prompt cuts in this PR. The purpose is to create a baseline that later PRs can cite with before/after numbers.

**Comment by alteixeira20** [comment]
Thanks both, this is exactly the kind of direction I was hoping this issue would attract.

@undergroundrap I think your proposed first PR shape is right: measurement-only, no prompt wording changes, no routing changes, no tool-selection changes, and focused tests around the accounting itself. Starting with a helper that reports section-level cost and largest contributors sounds like the safest baseline before touching behavior.

@kehansama thanks for the detailed notes and the disclosure. The memory-system point is especially useful. I agree Phase 0 should explicitly measure memory overhead instead of letting it disappear across tool definitions, retrieval prompts, and injected memory/context blocks.

I would probably keep “component budgets” as diagnostics/warnings at first rather than hard enforcement. Seeing that one area is consuming too much of the prompt budget is useful immediately, but truncation/enforcement can change behavior and should probably come after we have baseline numbers and eval coverage.

So I think the first implementation slice should stay narrow:

* assemble the same prompt sections the runtime would send;
* report char/token estimates per section;
* include memory/MCP/tool/integration/context sections explicitly;
* show largest contributors;
* avoid logging sensitive user content;
* add tests proving totals, percentages, sorting, and empty optional sections behave correctly;
* avoid prompt/routing/tool-selection behavior changes.

Then later PRs can use those numbers to justify MCP prose filtering, JIT tool guidance, context-pressure tests, and eventually budgets/enforcement if the data supports it.

That keeps this measurable instead of vibes-based.

**Comment by undergroundrap** [comment]
> Thanks both, this is exactly the kind of direction I was hoping this issue would attract.
> 
> [@undergroundrap](https://github.com/undergroundrap) I think your proposed first PR shape is right: measurement-only, no prompt wording changes, no routing changes, no tool-selection changes, and focused tests around the accounting itself. Starting with a helper that reports section-level cost and largest contributors sounds like the safest baseline before touching behavior.
> 
> [@kehansama](https://github.com/kehansama) thanks for the detailed notes and the disclosure. The memory-system point is especially useful. I agree Phase 0 should explicitly measure memory overhead instead of letting it disappear across tool definitions, retrieval prompts, and injected memory/context blocks.
> 
> I would probably keep “component budgets” as diagnostics/warnings at first rather than hard enforcement. Seeing that one area is consuming too much of the prompt budget is useful immediately, but truncation/enforcement can change behavior and should probably come after we have baseline numbers and eval coverage.
> 
> So I think the first implementation slice should stay narrow:
> 
> * assemble the same prompt sections the runtime would send;
> * report char/token estimates per section;
> * include memory/MCP/tool/integration/context sections explicitly;
> * show largest contributors;
> * avoid logging sensitive user content;
> * add tests proving totals, percentages, sorting, and empty optional sections behave correctly;
> * avoid prompt/routing/tool-selection behavior changes.
> 
> Then later PRs can use those numbers to justify MCP prose filtering, JIT tool guidance, context-pressure tests, and eventually budgets/enforcement if the data supports it.
> 
> That keeps this measurable instead of vibes-based.

Opened a focused Phase 0 PR here: #3059 

Scope is intentionally measurement-only:
- adds section-level prompt budget accounting;
- reports char/token estimates, percentages, and largest contributors;
- separates memory, MCP, builtin tool, integration, skill, and context sections;
- does not return raw prompt/user content in the report;
- does not change prompt wording, routing, tool selection, or enforcement.

Docker regression set is listed in the PR body.
